### DISCUSSION TITLE
How do I contribute to the editors?

### DISCUSSION THREAD

**Discussion by Heath123** [ONLYOFFICE/DocumentServer#1732] [Q&A]
Moved here from https://forum.onlyoffice.com/t/how-to-i-contribute-to-the-editors/1700

I’d like to fix some things that are annoying me in the editors (both web and desktop, as they seem to be the same codebase), but there’s very little information on how to contribute. https://www.onlyoffice.com/contribute.aspx doesn’t mention direct code contribution at all, only things like translations and plugins.

I can see that the desktop editor is just the web editor packaged in Chromium Embedded Framework, so it’s a bit like an Electron app. They both share the same code, which is in https://github.com/ONLYOFFICE/web-apps/.

This still leaves a lot of questions, though:

What is the best development workflow if I want to make changes? Should I test in the browser with the document server or the desktop app? If the desktop app, can I open the dev tools? How do I get an environment set up so that I can make changes? Do I just download the whole desktop or server source code, make changes to the web-apps submodule and re-run the build script for the whole thing? It seems to be set up for Ubuntu (I’m on Arch), and to use the master branch rather than the develop one - should I change this to the develop branch? I tried using the Snap image and doing `sudo mount --bind -o nodev,ro /home/heath/onlyoffice/web-apps/deploy/web-apps/ /var/lib/snapd/snap/onlyoffice-ds/current/var/www/onlyoffice/documentserver/web-apps/` as a hacky way to avoid doing the whole compilation and setup process, but this just seems to cause errors, probably due to a version mismatch.

**Comment by ShockwaveNN** [comment]
Hi, rather complicated question and our product is very complicated so we have no guides about `how to contribute, since it requires a lot of knowledge in different technologies.

So the best way to contribute - buy a license of our product )

> What is the best development workflow if I want to make changes?

It's described in this repo https://github.com/ONLYOFFICE/build_tools
But currently we do not support compiling on anything except Ubuntus

>  Should I test in the browser with the document server or the desktop app

DocumentServer is our base product, so checking changes in browser is simpler than DesktopEditors changes

> If the desktop app, can I open the dev tools

Run it with `--ascdesktop-support-debug-info` key

> How do I get an environment set up so that I can make changes?

Via build_tools

> Do I just download the whole desktop or server source code, make changes to the web-apps submodule and re-run the build script for the whole thing?

Something like this, but It very depend of your end goal

> It seems to be set up for Ubuntu (I’m on Arch), and to use the master branch rather than the develop one - should I change this to the develop branch?

Yes, we only support PR's from third party developers to develop

> I tried using the Snap image and doing sudo mount --bind -o nodev,ro /home/heath/onlyoffice/web-apps/deploy/web-apps/ /var/lib/snapd/snap/onlyoffice-ds/current/var/www/onlyoffice/documentserver/web-apps/ as a hacky way to avoid doing the whole compilation and setup process, but this just seems to cause errors, probably due to a version mismatch.

Never tried that, so not sure if it will work

**Comment by Heath123** [comment]
For anyone else here, I saw this, which I missed before: https://github.com/ONLYOFFICE/build_tools/blob/master/develop/README.md
(this looks like exactly what I was looking for but I haven't tried it yet)
It seems like the snap mess I was trying to do except actually supported and working
